spring-rabbit/src/main/java/org/springframework/amqp/rabbit/annotation/RabbitListenerAnnotationBeanPostProcessor.java

```diff
@@ -497,37 +494,110 @@ private String resolve(String value) {
+							resolveArguments(binding.arguments()));
 				}
 				else if (exchangeType.equals(ExchangeTypes.FANOUT)) {
 					exchange = new FanoutExchange(exchangeName,
 							resolveExpressionAsBoolean(bindingExchange.durable()),
-							resolveExpressionAsBoolean(bindingExchange.autoDelete()));
-					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, "", null);
+							resolveExpressionAsBoolean(bindingExchange.autoDelete()),
+							resolveArguments(bindingExchange.arguments()));
+					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, "",
+							resolveArguments(binding.arguments()));
 				}
 				else if (exchangeType.equals(ExchangeTypes.TOPIC)) {
 					exchange = new TopicExchange(exchangeName,
 							resolveExpressionAsBoolean(bindingExchange.durable()),
-							resolveExpressionAsBoolean(bindingExchange.autoDelete()));
-					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, resolvedKey, null);
+							resolveExpressionAsBoolean(bindingExchange.autoDelete()),
+							resolveArguments(bindingExchange.arguments()));
+					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, resolvedKey,
+							resolveArguments(binding.arguments()));
+				}
+				else if (exchangeType.equals(ExchangeTypes.HEADERS)) {
+					exchange = new HeadersExchange(exchangeName,
+							resolveExpressionAsBoolean(bindingExchange.durable()),
+							resolveExpressionAsBoolean(bindingExchange.autoDelete()),
+							resolveArguments(bindingExchange.arguments()));
+					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, resolvedKey,
+							resolveArguments(binding.arguments()));
 				}
 				else {
 					throw new BeanInitializationException("Unexpected exchange type: " + exchangeType);
 				}
-				((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment, exchange);
-				((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment, actualBinding);
+				((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment,
+						exchange);
+				((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment,
+						actualBinding);
 			}
 		}
 		return queues.toArray(new String[queues.size()]);
 	}
 
+	private Map<String, Object> resolveArguments(Argument[] arguments) {
+		Map<String, Object> map = new HashMap<String, Object>();
+		for (Argument arg : arguments) {
+			String key = resolveExpressionAsString(arg.name(), "name");
+			if (StringUtils.hasText(key)) {
+				Object value = resolveExpression(arg.value());
+				Object type = resolveExpression(arg.type());
+				Class<?> typeClass;
+				String typeName;
+				if (type instanceof Class) {
+					typeClass = (Class<?>) type;
+					typeName = typeClass.getName();
+				}
+				else {
+					Assert.isTrue(type instanceof String, "type must resolve to a Class or String");
```

Comment:
Similar message must be for the actual type, instead of just:
```
Caused by: java.lang.IllegalArgumentException: type must resolve to a Class or String
```
would be better to have this:
```
Assert.isTrue(type instanceof String, "type must resolve to a Class or String, but get [" +
							type.getClass().getName() + "]");
```
And therefore:
```
Caused by: java.lang.IllegalArgumentException: type must resolve to a Class or String, but get [java.util.Date]
```